drivers: flash: support for Nuvoton numaker series FMC

cyliangtw opened the pull request:
This PR is to add Nuvoton numaker series flash memory controller(FMC) of soc-flash.
It's verified on board: NuMaker-PFM-M467.

fabiobaltieri commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,290 @@
+
+static const struct flash_parameters flash_numaker_parameters = {
+#if DT_NODE_HAS_PROP(SOC_NV_FLASH_NODE, write_block_size)
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
```
Could you use `DT_PROP_OR` instead of the conditional?

cyliangtw replied:
Yes, DT_PROP_OR is better.

fabiobaltieri commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,290 @@
+	}
+
+move_exit:
+	SYS_LockReg();
+	irq_unlock(key);
```
is the order of this correct/important? Would have expected a symmetry between SYS_... and irq_...

cyliangtw replied:
Yes, the order is important and expected the pair of SYS_UnlockReg v.s SYS_LockReg, irq_lock v.s irq_unlock.

fabiobaltieri commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,290 @@
+	int block_size = flash_numaker_parameters.write_block_size;
+	int blocks = (len / flash_numaker_parameters.write_block_size);
+	uint8_t *pu8_data = (uint8_t *)data;
+#if DT_NODE_HAS_PROP(SOC_NV_FLASH_NODE, write_block_size)
+	uint8_t buffer[DT_PROP(SOC_NV_FLASH_NODE, write_block_size)];
+#else
+	uint8_t buffer[0x04];
+#endif
```
I see two points where you use this, maybe to do a #define with a default value and use it in both?

cyliangtw replied:
Thanks of your advice.
I define the value and use it in both.
Above modifications are in commit: 30d523233e9

fabiobaltieri commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,290 @@
+		}
+	}
+
+move_exit:
```
`move_exit`? What "move" means here? :-) The common labels for these are `out` or `done`...

cyliangtw replied:
I see, to change as "done".

cyliangtw commented on drivers/flash/Kconfig.numaker:
```diff
@@ -0,0 +1,16 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_FLASH_NUMAKER
+	bool "Nuvoton NUMAKER MCU flash driver"
```
Revise as "Nuvoton NuMaker MCU ...", is it right ?

fabiobaltieri replied:
Sure, I mean you are the one that came up with the capitalization of the brand :-) but yeah looks right now. You may want to update the comment on top of the file as well if you wanted to be picky.

cyliangtw commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,290 @@
+
+#define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
+
+extern int32_t FMC_EraseBank(uint32_t u32_bank_addr);
```
Yes, hal should be the better choice.
I notified the owner of exernal BSP/SDK to resolve it.
So, current alternative solution is to add this declaration to avoid compiler warning.

fabiobaltieri replied:
Could you open a PR for the HAL directly?

cyliangtw replied:
Yes, I could create one PR for HAL by Tomorrow. I will also try to ask external bsp/sdk .

fabiobaltieri commented:
@cyliangtw will you update the Nuvoton manifest to include https://github.com/zephyrproject-rtos/hal_nuvoton/pull/6 on this PR?

cyliangtw commented:
> @cyliangtw will you update the Nuvoton manifest to include [zephyrproject-rtos/hal_nuvoton#6](https://github.com/zephyrproject-rtos/hal_nuvoton/pull/6) on this PR?

Hi @fabiobaltieri ,
Yes, I did and verified it.
Thanks of your remind.

cyliangtw commented:
Hi @carlescufi ,
  This PR was updated to include https://github.com/zephyrproject-rtos/hal_nuvoton/pull/6.
If you are available, please have a look.

fabiobaltieri commented:
> Hi @carlescufi , This PR was updated to include [zephyrproject-rtos/hal_nuvoton#6](https://github.com/zephyrproject-rtos/hal_nuvoton/pull/6). If you are available, please have a look.

We need a review from the maintainer anyway. @de-nordic can you take a look?

de-nordic commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,281 @@
+	}
+
+	/* check for max value */
+	if ((offset + len) > (FMC_APROM_END - FMC_APROM_BASE)) {
```
You can still have very high value of `offset` and `len` that combined together will overflow and fit into the `FMC_APROM_END - FMC_APROM_BASE` allowing operation within not allowed range of flash offsets.

de-nordic commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,281 @@
+static bool flash_numaker_is_range_valid(off_t offset, size_t len)
+{
+	/* check for min value */
+	if ((offset < 0) || (len < 1)) {
```
`len < 1` is redundant check as `size_t` is unsigned and the only possible value to trigger this would be `len == 0`, something that callers of the function check before calling this funciton anywa.

cyliangtw replied:
Thanks and resolved

de-nordic commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,281 @@
+		addr += block_size;
+		blocks--;
+	}
+	if ((len % block_size) != 0) {
+		memcpy(buffer, (void *)addr, block_size);
+		memcpy(buffer, (void *)pu8_data, len / block_size);
+		if (flash_numaker_block_write(addr, buffer, block_size)) {
+			rc = -EIO;
+			goto done;
+		}
+	}
```
Writes unaligned to `write_block_size` are not allowed, write callback should not correct user:
https://docs.zephyrproject.org/apidoc/latest/group__flash__interface.html#ga76d7880cc5e18ca40237736d3bd94324

cyliangtw replied:
Revise the flash_numaker_write() based on your advice & flash_write spec.
Update at commit  e8777cd8a83f

de-nordic commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	/* check for max value */
+	if ((offset + len) < 0) {
```
This will not work on a principle how casting in C works. All the parameters for the equation will be promoted to the highest unsigned type in equation, which in this case would be `size_t`, so the result of this equation is basically unsigned.
I do not think it is possible to weasel out, this way, from checking `len` and `offset` against allowed range.

cyliangtw replied:
Yes, it will casting as unsigned int in C.
I changed to use int64_t to cover it in 32 bit Cortex-M platform.
Update at commit  bcec8a3355338c.

cyliangtw replied:
Rebase to apply  #60551 to resolve warning `API_DEFINE: do not specify a non-Zephyr API for libc`.

cyliangtw replied:
Fixed at commit 1e4b14fd571

de-nordic commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,282 @@
+	}
+
+	/* check for max value */
+	if (((int64_t) offset + len) > ((int64_t) FMC_APROM_END - FMC_APROM_BASE)) {
+		return false;
+	}
```
Casting up works fine as long as you cast up to processor native word. Above you just ask it to engage in extra work as, probably, dropping registers to stack to make place for 64bit calculation (32 bit addition == two/three 32bit registers, 64bit addition 4/6 32bit registers), and making one more addition.
This is not the first time I can see that so I will just ask: what is wrong with something like
```
if (offset > (FMC_APROM_END - FMC_APROM_BASE) || len > (FMC_APROM_END - FMC_APROM_BASE)  || (len + offset) >  (FMC_APROM_END - FMC_APROM_BASE) ) {
}
```
?
Here is assembly for non optimized code:
```
000098f4 <decide_64>:
{
    98f4:       e92d 0fb0       stmdb   sp!, {r4, r5, r7, r8, r9, sl, fp}
    98f8:       b087            sub     sp, #28
    98fa:       af00            add     r7, sp, #0
    98fc:       6178            str     r0, [r7, #20]
    98fe:       6139            str     r1, [r7, #16]
    9900:       60fa            str     r2, [r7, #12]
        return ((int64_t)(s) + (int64_t)(o)) < total;
    9902:       697b            ldr     r3, [r7, #20]
    9904:       2200            movs    r2, #0
    9906:       603b            str     r3, [r7, #0]
    9908:       607a            str     r2, [r7, #4]
    990a:       693b            ldr     r3, [r7, #16]
    990c:       17da            asrs    r2, r3, #31
    990e:       469a            mov     sl, r3
    9910:       4693            mov     fp, r2
    9912:       e9d7 2300       ldrd    r2, r3, [r7]
    9916:       4611            mov     r1, r2
    9918:       eb11 040a       adds.w  r4, r1, sl
    991c:       eb43 050b       adc.w   r5, r3, fp
    9920:       68fb            ldr     r3, [r7, #12]
    9922:       2200            movs    r2, #0
    9924:       4698            mov     r8, r3
    9926:       4691            mov     r9, r2
    9928:       4544            cmp     r4, r8
    992a:       eb75 0309       sbcs.w  r3, r5, r9
    992e:       bfb4            ite     lt
    9930:       2301            movlt   r3, #1
    9932:       2300            movge   r3, #0
    9934:       b2db            uxtb    r3, r3
}
    9936:       4618            mov     r0, r3
    9938:       371c            adds    r7, #28
    993a:       46bd            mov     sp, r7
    993c:       e8bd 0fb0       ldmia.w sp!, {r4, r5, r7, r8, r9, sl, fp}
    9940:       4770
```
vs
```
00009942 <decide_if>:
{
    9942:       b480            push    {r7}
    9944:       b085            sub     sp, #20
    9946:       af00            add     r7, sp, #0
    9948:       60f8            str     r0, [r7, #12]
    994a:       60b9            str     r1, [r7, #8]
    994c:       607a            str     r2, [r7, #4]
        return (o < total && s < total && (s + o) < total);
    994e:       68bb            ldr     r3, [r7, #8]
    9950:       687a            ldr     r2, [r7, #4]
    9952:       429a            cmp     r2, r3
    9954:       d90b            bls.n   996e <decide_if+0x2c>
    9956:       68fa            ldr     r2, [r7, #12]
    9958:       687b            ldr     r3, [r7, #4]
    995a:       429a            cmp     r2, r3
    995c:       d207            bcs.n   996e <decide_if+0x2c>
    995e:       68ba            ldr     r2, [r7, #8]
    9960:       68fb            ldr     r3, [r7, #12]
    9962:       4413            add     r3, r2
    9964:       687a            ldr     r2, [r7, #4]
    9966:       429a            cmp     r2, r3
    9968:       d901            bls.n   996e <decide_if+0x2c>
    996a:       2301            movs    r3, #1
    996c:       e000            b.n     9970 <decide_if+0x2e>
    996e:       2300            movs    r3, #0
    9970:       f003 0301       and.w   r3, r3, #1
    9974:       b2db            uxtb    r3, r3
}
    9976:       4618            mov     r0, r3
    9978:       3714            adds    r7, #20
    997a:       46bd            mov     sp, r7
    997c:       bc80            pop     {r7}
    997e:       4770            bx      lr
```
Just saving/restoring registers on entry/exit takes extra twelve word transfers.

cyliangtw replied:
My viewpoint is to keep the result of "offset+len" and not missing any bit caused by overflow.
In fact, your suggestion is working for this case and got better code optimization.

cyliangtw replied:
I applied your suggestion for better performance and is enough for this validation.
Update at commit 5baf96a32bb7.
Thanks.

de-nordic reviewed (dismissed):
Some comment left, otherwise OK.

cyliangtw committed:
drivers: flash: support for Nuvoton numaker series FMC

Add Nuvoton numaker series flash memory controller(FMC) with erase,
 read & write features of soc-flash. Also update Nuvoton manifest
 to include zephyrproject-rtos/hal_nuvoton#6.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw commented on drivers/flash/soc_flash_numaker.c:
```diff
@@ -0,0 +1,284 @@
+	}
+
+	/* check for max value */
+	if (offset > aprom_size || len > aprom_size || (offset + len) > aprom_size) {
```
Yes, it's really the best way and will never suffer overflow.

cyliangtw replied:
Update this best way at commit b218f2e94b

de-nordic reviewed (approved):
Looks OK. Thanks for contributing.

cyliangtw commented:
> @cyliangtw I guess that here entire platform is added, so I just want to ask if there will be addition to release notes regarding the added support?

@de-nordic In fact, I2C/PWM/USB/EMAC/CAN-FD drivers were done locally in March and 
 I will continue to send PR one by one. Is the release notes to describe the entire platform or just only one specific driver support ? Is there any reference ?

cyliangtw commented:
> > @cyliangtw I guess that here entire platform is added, so I just want to ask if there will be addition to release notes regarding the added support?
> 
> @de-nordic In fact, I2C/PWM/USB/EMAC/CAN-FD drivers were done locally in March and I will continue to send PR one by one. Is the release notes to describe the entire platform or just only one specific driver support ? Is there any reference ?

@de-nordic  I see and I will add notes in [doc/releases/release-notes-3.5.rst](https://github.com/zephyrproject-rtos/zephyr/pull/60660/files#diff-0cc5759a7c2121d6afcbb4540538a7a3a08e135593a1493b01811981df4dc126)

cyliangtw commented:
@de-nordic, 
release notes in PR #60667
